src/authhandler.cpp

```diff
@@ -51,61 +51,36 @@ const QString AUTH_PATH             ("AuthPath");
 const QString TOKEN_PATH            ("TokenPath");
 const QString REDIRECT_URI          ("RedirectUri");
 const QString HOST                  ("Host");
-const QString SLASH                 ("/");
-const QString AUTH                  ("auth");
-const QString AUTH_METHOD           ("method");
-const QString MECHANISM             ("mechanism");
 
-AuthHandler::AuthHandler(Accounts::Manager *manager, const quint32 accountId, const QString &accountService, QObject *parent)
+AuthHandler::AuthHandler(Accounts::AccountService *service, QObject *parent)
     : QObject(parent)
-    , mAccountManager(manager)
-    , mAccount(manager->account(accountId))
-    , m_accountService(accountService)
+    , mAccountService(service)
 {
 }
 
 bool AuthHandler::init()
 {
     FUNCTION_CALL_TRACE(lcCalDavTrace);
 
-    if (mAccount == NULL) {
-        qCDebug(lcCalDav) << "Invalid account";
+    if (mAccountService == NULL) {
+        qCDebug(lcCalDav) << "Invalid account service";
         return false;
     }
-
-    QVariant val = QVariant::String;
-
-    Accounts::Service srv = mAccountManager->service(m_accountService);
-    if (!srv.isValid()) {
-        qCWarning(lcCalDav) << "Cannot select service:" << m_accountService;
+    const Accounts::AuthData &auth = mAccountService->authData();
+    if (auth.credentialsId() == 0) {
+        qCWarning(lcCalDav) << "Cannot authenticate, no credentials stored for service:" << mAccountService->service().name();
         return false;
     }
-    mAccount->selectService(srv);
-    if (!mAccount->enabled()) {
-        qCWarning(lcCalDav) << "Service:" << m_accountService << "is not enabled for account:" << mAccount->id();
-        return false;
-    }
-
-    mAccount->value(AUTH + SLASH + AUTH_METHOD, val);
-    mMethod = val.toString();
-    mAccount->value(AUTH + SLASH + MECHANISM, val);
-    mMechanism = val.toString();
-    qint32 cId = mAccount->credentialsId();
-    mAccount->selectService(Accounts::Service());
 
-    if (cId == 0) {
-        qCWarning(lcCalDav) << "Cannot authenticate, no credentials stored for service:" << m_accountService;
-        return false;
-    }
-    mIdentity = Identity::existingIdentity(cId);
+    mIdentity = Identity::existingIdentity(auth.credentialsId());
     if (!mIdentity) {
-        qCWarning(lcCalDav) << "Cannot get existing identity for credentials:" << cId;
+        qCWarning(lcCalDav) << "Cannot get existing identity for credentials:" << auth.credentialsId();
         return false;
     }
 
-    mSession = mIdentity->createSession(mMethod.toLatin1());
+    mSession = mIdentity->createSession(auth.method().toLatin1());
```

Comment:
Um, this returns a QPointer<> but it's assigned to plain pointer and also leaks memory?

Reply:
According to the documentation, the session returned by an identity is owned by the identity. It should not be freed. I think there is no memory leak by just doing nothing with the pointer on return.

Reply:
Ah true. Since identity is owned by this it should all go better now. Might still not be bad idea to store the value with the QPointer that's gotten here, but yea.